aiida/tools/graph/graph_traversers.py

```diff
@@ -0,0 +1,287 @@
+    """
+    This function will return the set of all nodes that can be connected
+    to a list of initial nodes through any sequence of specified links.
+    Optionally, it may also return the links that connect these nodes.
+
+    :type starting_pks: list or tuple
+    :param starting_pks: The set of (valid) pks for the starting nodes (also accepts sets).
+
+    :type max_iterations: int or None
+    :param max_iterations:
+        The number of iterations to apply the set of rules (a value of 'None' will
+        iterate until no new nodes are added).
+
+    :type get_links:  bool
+    :param get_links:
+        Pass True to also return the links between all nodes (found + initial).
+
+    :type links_forward: aiida.common.links.LinkType
+    :param links_forward:
+        List with all the links that should be traversed in the forward direction.
+
+    :type links_backward: aiida.common.links.LinkType
+    :param links_backward:
+        List with all the links that should be traversed in the backward direction.
+    """
+    # pylint: disable=too-many-locals,too-many-statements,too-many-branches
+    from aiida import orm
+    from aiida.tools.graph.age_entities import Basket
+    from aiida.tools.graph.age_rules import UpdateRule, RuleSequence, RuleSaveWalkers, RuleSetWalkers
+    from aiida.common import exceptions
+
+    if max_iterations is None:
+        max_iterations = inf
+    elif not (isinstance(max_iterations, int) or max_iterations is inf):
+        raise TypeError('Max_iterations has to be an integer or infinity')
+
+    linktype_list = []
+    for linktype in links_forward:
+        if not isinstance(linktype, LinkType):
+            raise TypeError('links_forward should contain links, but one of them is: {}'.format(type(linktype)))
+        linktype_list.append(linktype.value)
+    filters_forwards = {'type': {'in': linktype_list}}
+
+    linktype_list = []
+    for linktype in links_backward:
+        if not isinstance(linktype, LinkType):
+            raise TypeError('links_backward should contain links, but one of them is: {}'.format(type(linktype)))
+        linktype_list.append(linktype.value)
+    filters_backwards = {'type': {'in': linktype_list}}
+
+    if not isinstance(starting_pks, (list, set, tuple)):
+        raise TypeError('starting_pks must be of type list, set or tuple\ninstead, it is {}'.format(type(starting_pks)))
+
+    if not starting_pks:
+        if get_links:
+            output = {'nodes': set(), 'links': set()}
+        else:
+            output = {'nodes': set(), 'links': None}
+        return output
+
+    operational_set = set()
+    for single_pk in starting_pks:
+        if not isinstance(single_pk, int):
+            raise TypeError(
+                'element provided in starting_pks cannot be pk because it is of type: {}'.format(type(single_pk))
+            )
+        operational_set.add(single_pk)
+
+    query_nodes = orm.QueryBuilder()
```

Comment:
This might be a costly check. Do we really need it? Would ignoring non-existing pks be such a bad idea?

Reply:
I'm usually wary of functions that simply decide that there is nothing to be done with your instructions without actually letting you know about this (basically helping to hide typos and errors in scripts). Maybe a compromise solution would be to add an input keyword 'ignore_absent' that is checked before doing the querying? (I would personally default it to `True` but I have less strong feelings about that)

Reply:
Actually on second thought, the check is not that costly as it will only fetch the id's that are in the starting set, which typically is not that big. Secondly, I now actually agree with you that it is better not to silently swallow non-existing ids. So let's keep it like it was (you can remove the `ignore_absent` again, sorry .... :)

Reply:
No probs, already done.